src/clock/index.js

```diff
@@ -122,13 +122,16 @@ const contains = async (events, a, b) => {
   if (a.toString() === b.toString()) return true
   const [{ value: aevent }, { value: bevent }] = await Promise.all([events.get(a), events.get(b)])
   const links = [...aevent.parents]
+  const seen = new Set()
   while (links.length) {
     const link = links.shift()
     if (!link) break
     if (link.toString() === b.toString()) return true
     // if any of b's parents are this link, then b cannot exist in any of the
     // tree below, since that would create a cycle.
     if (bevent.parents.some(p => link.toString() === p.toString())) continue
+    if (seen.has(link.toString())) continue
+    seen.add(link.toString())
     const { value: event } = await events.get(link)
```

Comment:
The reasoning here is that if you have a tree that branches and then merges, this prevents the program from crawling the pre-merge history again. Without this, each merge in the tree makes all history below it traversed N times (where N is the parents.length in the merge).